runtime/runtime/src/prefetch.rs

```diff
@@ -86,23 +124,38 @@ impl TriePrefetcher {
         &mut self,
         transactions: &[SignedTransaction],
     ) -> Result<(), ()> {
-        for t in transactions {
-            let account_id = t.transaction.signer_id.clone();
-            let trie_key = TrieKey::Account { account_id };
-            self.prefetch_trie_key(trie_key)?;
+        if self.prefetch_api.enable_receipt_prefetching {
+            for t in transactions {
+                let account_id = t.transaction.signer_id.clone();
+                let trie_key = TrieKey::Account { account_id };
+                self.prefetch_trie_key(trie_key)?;
 
-            let trie_key = TrieKey::AccessKey {
-                account_id: t.transaction.signer_id.clone(),
-                public_key: t.transaction.public_key.clone(),
-            };
-            self.prefetch_trie_key(trie_key)?;
+                let trie_key = TrieKey::AccessKey {
+                    account_id: t.transaction.signer_id.clone(),
+                    public_key: t.transaction.public_key.clone(),
+                };
+                self.prefetch_trie_key(trie_key)?;
+            }
         }
         Ok(())
     }
 
-    /// Removes all queue up prefetch requests.
+    /// Removes all queued up prefetch requests and staged data.
+    ///
+    /// Note that IO threads currently prefetching a trie key might insert
+    /// additional trie nodes afterwards. Therefore the staging area is not
+    /// reliably empty after resetting.
+    /// This is okay-ish. Resetting between processed chunks does not have to
+    /// be perfect, as long as we remove each unclaimed value eventually. Doing
+    /// it one chunk later is also okay.
+    ///
+    /// TODO: In the presence of forks, multiple chunks of a shard can be processed
+    /// at the same time. They share a prefetcher, so they will clean each others
+    /// data. Handling this is a bit more involved. Failing to do so makes prefetching
+    /// less effective in those cases but crucially nothing breaks.
     pub(crate) fn clear(&self) {
-        self.prefetch_api.clear();
+        self.prefetch_api.clear_queue();
+        self.prefetch_api.clear_data();
     }
 
     fn prefetch_trie_key(&self, trie_key: TrieKey) -> Result<(), ()> {
```

Comment:
and a metric here -- on how many prefetches we're requesting (and how many of them are failing due to queue full)